File: lintcode/add_binary/cpp/add_binary.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/add-binary
@Language: C++
@Datetime: 17-02-15 18:17
*/

#include <iterator>

class Solution {
public:
    /**
     * @param a a number
     * @param b a number
     * @return the result
     */
    std::string addBinary(std::string& left, std::string& right)
    {
        static const int BASE{ 2 };
        
        std::string result;
        
        auto i = left.size();
        auto j = right.size();
        int carry{ 0 };
        
        while (i > 0 || j > 0)
        {
            const auto leftVal  = i > 0 ? left[i-- - 1] - '0' : 0;
            const auto rightVal = j > 0 ? right[j-- - 1] - '0' : 0;
            const auto sum      = leftVal + rightVal + carry;
            
            result += sum % BASE + '0';
            carry   = sum / BASE;
        }
        
        if (carry) {
            result += carry % BASE + '0';
        }
        
        return {
            std::reverse_iterator<std::string::const_iterator>{ result.cend() },
            std::reverse_iterator<std::string::const_iterator>{ result.cbegin() }
        };
    }
};

File: lintcode/easy/add_binary/py/add_binary.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/add-binary
@Language: Python
@Datetime: 17-02-15 14:35
'''

class Solution:
    # @param {string} a a number
    # @param {string} b a number
    # @return {string} the result
    def addBinary(self, a, b):
        def xor(a, b):
            import operator
                
            left  = map(int, a.zfill(len(b)))
            right = map(int, b.zfill(len(a)))
                
            return ''.join(str(bit) for bit in map(operator.xor, left, right))
            
        def and_(a, b):
            import operator
                
            left  = map(int, a.zfill(len(b)))
            right = map(int, b.zfill(len(a)))
                
            return ''.join(str(bit) for bit in map(operator.and_, left, right))
                
        def lsh(a, b):
            return a + '0' * int(b)
            
        def helper(a, b):
            if len(b) == 0:
                return a if len(a) > 0 else '0'
                
            total = xor(a, b)
            carry = lsh(and_(a, b), 1)
                    
            return helper(total.lstrip('0'), carry.lstrip('0'))
            
        return helper(a, b)